include/nbl/builtin/glsl/utils/swapchain_transform.glsl

```diff
@@ -0,0 +1,146 @@
+        return ivec2(lastTexel.y - coord.y, coord.x);
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_ROTATE_180:
+        return ivec2(lastTexel) - coord;
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_ROTATE_270:
+        return ivec2(coord.y, lastTexel.x - coord.x);
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR:
+        return ivec2(lastTexel.x - coord.x, coord.y);
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR_ROTATE_90:
+        return lastTexel - coord.yx;
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR_ROTATE_180:
+        return ivec2(coord.x, lastTexel.y - coord.y);
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR_ROTATE_270:
+        return coord.yx;
+    default:
+        return ivec2(0);
+    }
+}
+
+// If using `imageStore` to write *directly to the swapchain*, you may use this function to 
+// transform the UVs before rendering with them (e.g. path tracer that will use UVs to fire rays, gather). 
+// Remember to only use one transform in the pipeline!
+ivec2 nbl_glsl_swapchain_transform_preTransform(in uint swapchainTransform, in ivec2 coord, in ivec2 screenSize) {
+    return nbl_glsl_swapchain_transform_backward(swapchainTransform, coord, screenSize - 1);
+}
+
+// If using `imageStore` to write *directly to the swapchain*, you may use this function to 
+// transform the coordinates written to the swapchain (the ones used in imageStore, scatter).
+// Remember to only use one transform in the pipeline!
+ivec2 nbl_glsl_swapchain_transform_postTransform(in uint swapchainTransform, in ivec2 coord, in ivec2 screenSize) {
+    return nbl_glsl_swapchain_transform_forward(swapchainTransform, coord, screenSize - 1);
+}
+
+// If rendering directly to the swapchain, use the transformed extents rather than the actual
+// extents for aspect ratio calculations, in order to be consistent with 90º/270º rotations.
+ivec2 nbl_glsl_swapchain_transform_preTransformExtents(in uint swapchainTransform, in ivec2 screenSize) {
+    switch (swapchainTransform) 
+    {
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_IDENTITY:
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR:
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR_ROTATE_180:
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_ROTATE_180:
+        return screenSize;
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_ROTATE_90:
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_ROTATE_270:
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR_ROTATE_90:
+    case NBL_GLSL_SWAPCHAIN_TRANSFORM_E_HORIZONTAL_MIRROR_ROTATE_270:
+        return screenSize.yx;
+    default:
+        return ivec2(0);
+    }
+}
+
+// TODO: nbl_glsl_swapchain_transform_transformedDerivatives is untested
+// If rendering directly to the swapchain, ddx/ddy operations may be incorrect due to the swapchain
+// transform. Use this helper function to transform the ddx or ddy accordingly.
+mat2 nbl_glsl_swapchain_transform_transformedDerivatives(in uint swapchainTransform, in mat2 ddxDdy) {
```

Comment:
Good description/comment 👍 
just a little change in the name because derivatives can be applied to floats too rather than vec2's:
`nbl_glsl_swapchain_transform_transformedDerivatives2D`
Here is the table I made:

| **Surface Transform**            | **new x derivative** | **new y derivative** |
|:--------------------------------:|:--------------------:|:--------------------:|
| **IDENTITY**                     | dFdx                 | dFdy                 |
| **HORIZONTAL_MIRROR**            | -dFdx                | dFdy                 |
| **HORIZONTAL_MIRROR_ROTATE_180** | dFdx                 | -dFdy                |
| **ROTATE_180**                   | -dFdx                | -dFdy                |
| **ROTATE_90**                    | dFdy                 | -dFdx                |
| **ROTATE_270**                   | -dFdy                | dFdx                 |
| **HORIZONTAL_MIRROR_ROTATE_90**  | dFdy                 | dFdx                 |
| **HORIZONTAL_MIRROR_ROTATE_270** | -dFdy                | -dFdx                |


Because we're 2D, dFdx coressponds to the first element in the `mat2 ddxDdy` which is a `vec2` and similar for the dFfy and second element (verify @devshgraphicsprogramming)

Reply:
in GLSL when you do `mat2(vec2,vec2)` you are initializing subsequent columns

Reply:
You could just make a full set of overloads
```glsl
vec2 nbl_glsl_swapchain_transform_transformedDerivatives(in uint swapchainTransform, in vec2 DFDScreen);
mat2 nbl_glsl_swapchain_transform_transformedDerivatives(in uint swapchainTransform, in mat2 DFDScreen);
mat2x3 nbl_glsl_swapchain_transform_transformedDerivatives(in uint swapchainTransform, in mat2x3 DFDScreen);
mat2x4 nbl_glsl_swapchain_transform_transformedDerivatives(in uint swapchainTransform, in mat2x4 DFDScreen);
```

perhaps implement using an Xmacro `#include`